### ISSUE TITLE
Spurious type error in nested function

### ISSUE THREAD

**Issue by NeilGirdhar** [python/mypy#9325]
```python
from typing import Optional, Mapping


def f(x: Optional[Mapping[str, int]]) -> int:
    if x is None:
        return 0
    def g() -> int:
        return x['a']
    return g()
```
gives
```
a.py:8: error: Value of type "Optional[Mapping[str, int]]" is not indexable
```

**Comment by hauntsaninja** [comment]
Duplicate of https://github.com/python/mypy/issues/2608, https://github.com/python/mypy/issues/9197 and others

**Comment by NeilGirdhar** [comment]
Thanks, and sorry for the duplicate!

**Comment by hauntsaninja** [comment]
All good, thanks for reporting! :-) This one can be tricky. Also an easy workaround for your case is aliasing the variable, eg, if you insert a `y = x` after the return and then use `y` everywhere, it should work.
